### ISSUE TITLE
Support table driven tests

### ISSUE THREAD

**Issue by helmbold** [kotlintest/kotlintest#5]
Some form of parametrized tests like ScalaTests [table driven](http://www.scalatest.org/user_guide/table_driven_property_checks) test would be nice.

**Comment by sksamuel** [comment]
Seems similar to scala check.

**Comment by helmbold** [comment]
The linked page is a subpage of [property based testing](http://www.scalatest.org/user_guide/property_based_testing), so, yes, I'd say it is similar. The tables are just my favourite feature.

**Comment by sksamuel** [comment]
I'll add something like that in the next few days.

**Comment by helmbold** [comment]
Cool!

**Comment by sksamuel** [comment]
I'm focusing on property based testing first, as that will then be used by the tables. Because Kotlin doesn't have implicits and because the runtime type parameters of functions are erased, it's going to be more verbose than Scala. 

Something like this atm:

```kotlin
class PropSpecTest : PropSpec() {

  init {
    property("startsWith") {
      forAll(Generator.string(), Generator.string(), { a, b ->
        (a + b).startsWith(a)
      })
    }
  }
}
```

**Comment by helmbold** [comment]
I have not thought about it in detail, but maybe you could use extension methods for more concise code here? In any case, I think it is worth to think some time about the implementation of property based testing.

**Comment by sksamuel** [comment]
What's your thinking re:extension methods? Any brainstorming you can put here?

**Comment by helmbold** [comment]
My first idea was do write something like this:

```kotlin
init {
    "starts with".forAll {
        ...
    }
}
```

But after thinking about it, I think the word `property` is helpful to express the tests. So, implicit conversions might not be helpful for this API (at least I have no idea how).

What do you think about the following builder style? 

```kotlin
property("...").forAll(Generator.string(), Generator.string()) { a, b ->
    (a + b).startsWith(a)
}
```

It would reduce one level of nesting and the description of a property could actually be seen as a builder.

Maybe the generator could be looked up by the `forAll` implementation with reified generics:

```kotlin 
inline fun <reified T> showTypeParameter(noinline f: (T, T) -> T) {
    println(T::class)
}
```

My example above could then be written like this:

```kotlin
property("...").forAll { a, b ->
    (a + b).startsWith(a)
}
```

In your example you extend `PropSpec`; it would be nice to be able to combine usual testing styles and property based testing. It is very handy to be able to test some aspects with property based testing and some with usual testing in the same test class.

**Comment by sksamuel** [comment]
Thanks for the feedback.

The builder style seems quite nice:

```kotlin
property("...").forAll(Generator.string(), Generator.string()) { a, b ->
    (a + b).startsWith(a)
}
```

I wasn't aware reified generics was in Kotlin, so I shall check that out. Assuming I can get a handle on the function type parameters, then it will be easy to infer generators for them, although if one is specified, all will need to be specified (not a big deal) (This is where Scala's implicits are a gem. I hope they add some typeclass support to Kotlin).

And property testing can be included in all styles. In fact there's no real reason to have the different classes at all, could just be a `KotlinTestSuperClass` that has them all in.

**Comment by sksamuel** [comment]
Ok that works, so you could do this:

```kotlin
property("startsWith").forAll { a: String, b: String ->
    (a + b).startsWith(a)
}
```

or 

```kotlin
property("startsWith").forAll<String,String> { a, b ->
    (a + b).startsWith(a)
}
```

Obviously, would be a default generator for all basic types.

**Comment by sksamuel** [comment]
An example of what we have thus far:
https://github.com/kotlintest/kotlintest/blob/master/src/test/kotlin/io/kotlintest/properties/PropSpecTest.kt

**Comment by helmbold** [comment]
It looks pretty good!

Custom generators could be set via a method or provided by an overridden property. Example:

```kotlin
init {
    addGenerator(MyIntGenerator) // using reified generics to get the actual type of the generator
    // or set the type explicitly
    putGenerator(Int::class, MyIntGenerator)
    // or
    put<Int>(MyIntGenerator)
}
```
`add...` is perhaps a bit misleading, since it would actually override the default IntGenerator.

**Comment by sksamuel** [comment]
You could allow registration of defaults, and then you wouldn't need to specify.
Definitely need to still be able to specify on the function for when you want two different string implementations in the same function for example.

**Comment by sksamuel** [comment]
Kotlin doesn't have tuples so its going to be a lot more verbose creating Tables. One way would be to have a Table that accepts a vararg of some type T, where T contains the fields you want to use. Perhaps some types could be provided similar to Scala's Tuple2, Tuple3, etc.

**Comment by helmbold** [comment]
The problem with vararg is that all elements must be of the same type. So I think, that it is no option. Maybe it would be better to have something like Table1<A>, Table2<A, B>, Table3<A, B, C> and so on. The limit should obviously be Table22, since that is the one and only true parameter limit as proofed by Scala ;-)

**Comment by sksamuel** [comment]
What I mean is the types of the tuples/rows inside the table.

**Comment by helmbold** [comment]
Yes, I mean the same. But in the end, the table itself have to carry this same type. A table could consist of several tuples or rows (I think that is better in the context of tables).

**Comment by sksamuel** [comment]
In Scala though you have typed tuples which Kotlin doesn't have. So we have
to create our own "row" object.

**Comment by helmbold** [comment]
The Kotlin devs [dropped tuples](http://blog.jetbrains.com/kotlin/migrating-tuples/), because they say that you should better use data classes. Having seen some misuses of meaningless tuples, I can agree with it. Creating 22 ;-) Row classes is not that bad, I think. 22 lines should be enough in Kotlin.

**Comment by sksamuel** [comment]
I do miss tuples though I think it was it a wrong decision to leave out.
But yes 20 or so Row classes.

**Comment by russel** [comment]
For PropSpec I would expect there to be a default of not having to specify generators at all, but having the ability to specify generators for special cases. So something like:
```kotlin
class PropSpecTest : PropSpec() {
  init {
    property("startsWith") {
      forAll { a:String, b:String ->
        (a + b).startsWith(a)
      })
    }
  }
}
```

**Comment by sksamuel** [comment]
You can do basically that and it uses reflection to work out the
generators. I'll be adding more docs when it's properly released. Any
requirements let me know.

**Comment by russel** [comment]
Is there a snapshot artefact I can use prior to the next release? I appreciate the last resort is clone the repository and build myself, but I was wondering if there were snapshots in a Gradle accessible place.

**Comment by sksamuel** [comment]
No not at the moment. It's something I need to setup.

**Comment by sksamuel** [comment]
Saying that the generator work in progress is in 1.1.6 just not advertised
yet in docs.

**Comment by russel** [comment]
`PropSpec` seems not to be in 1.1.6. I'll compile Git master.

**Comment by sksamuel** [comment]
Oh ok sorry.

**Comment by russel** [comment]
I compiled up Git master, discovered it's called `PropertyTesting` and no `PropSpec` from the examples (yay for examples ;-) ). Everything compiles but on execution I am missing:

com/sksamuel/koors/Random

I guess this is a transitive dependency I do not have?

**Comment by sksamuel** [comment]
Hmm should be on central. Otherwise travis wouldn't compile.

**Comment by russel** [comment]
OK, I misspelled the dependency :-(

Property-based tests now starting but it hangs there.

**Comment by sksamuel** [comment]
I'm on holiday atm but once back I intend to get 1.2.0 released with the
prop testing offficial and the other outstanding requests.

**Comment by russel** [comment]
OK, I'll stop hassling you for now. Have fun.

**Comment by sksamuel** [comment]
Haha no hassle :)

**Comment by sksamuel** [comment]
Bumping this to 1.3 so I can release 1.2 with the grouping/config stuff.

**Comment by sksamuel** [comment]
I will revisit this in the next few days @helmbold. I guess the best we can do is create tables using Row objects, where we make all the different Row1...RowN variations. And that's how you would build up your tables, unless you have a better suggestion.

**Comment by helmbold** [comment]
I think classes for Row1 ... Row22 is just fine. Maybe the first row could be used for headings to be able to generate more meaningful output / error messages.

I imagine code like this:

```kotlin
table(
    row("price", "amount", "tax rate", "expected result"),
    row(9.99, 2, 0.2, 23.98),
    row(45.00, 1, 0.1, 49,45)
)
```
`row` and `table` should be some kind of factory methods, so that the user isn't required to type `Row4` and the like.

**Comment by sksamuel** [comment]
Funny, I came up with exactly that when playing around at lunchtime (the capitalised version). But I quite like your lower case one.

Then its easy to add

`fun <A,B> foreach(table:Table<A,B>, fn: (A,B) -> Unit): Unit = { }`

**Comment by helmbold** [comment]
I'm not sure, but maybe the header row could have a special name, to make the meaning obvious:

```
table(
    header("price", "amount", "tax rate", "expected result"),
    row(9.99, 2, 0.2, 23.98),
    row(45.00, 1, 0.1, 49,45)
)
```
Otherwise it looks somehow magic that the first `row` describers the header.

**Comment by sksamuel** [comment]
I like that idea also.

**Comment by russel** [comment]
Scala doesn't bother with a row class at all, a table is a sequence of tuples, the first defines the names and the others (all the same length) provide the data:
```
table(
    ("price", "amount", "tax rate", "expected result"),
    (9.99, 2, 0.2, 23.98),
    (45.00, 1, 0.1, 49,45)
)
```
The signle biggest problem is you cannot have a tralining comma on the last entry, this is a real pain in the proverbials.

**Comment by helmbold** [comment]
@russel But Kotlin doesn't have tuples. Having seen some cases of tuple abuse in Scala (e. g. programmer was too lazy to find a class name), I'm not that sad.

**Comment by sksamuel** [comment]
I hate the fact they missed out tuples. I use them a lot in Scala. They can be misused but so can functions, and most other language features. It's very handy when you want to just have a quick pair passed about, and Pair<> is just too noisy. I'd have been happy if they'd have limited tuples to say 2 or 3. After that you probably should go for a data class.

**Comment by russel** [comment]
But Kotlin does have lists, so there is the option of doing that rather than inventing a variadic data class. The table would have to check the length of the row anyway.

**Comment by sksamuel** [comment]
The issue with lists is that they are singly typed. A HList would be fine, but Kotlin doesn't have that. A data class can be type parameterised so the type of the values in the functions can be inferred.

**Comment by russel** [comment]
In the end, whatever works with minimum overhead in the code works for me.

**Comment by russel** [comment]
It might also be worth adding that now I can properly do property testing, I don't have much of a need for data table testing.

**Comment by helmbold** [comment]
To come back to tables: After thinking a while about my approaches with the factory object (uppercase for users) and top-level factory functions (lowercase for users), I prefer the latter. I think it is more idiomatic Kotlin to leave out an otherwise useless object. 

Compare the following options:

```kotlin
object Row {
    operator fun <A, B> invoke(a: A, b: B) = Row2(a, b)
}
```
vs.

```kotlin
fun <A, B> row(a: A, b: B) = Row2(a, b)
```

I think it is obvious what is easier to read and with less clutter.

**Comment by sksamuel** [comment]
I've added the first bash, obviously open to change on feedback

**Comment by helmbold** [comment]
one more type parameter than Scala could handle ;-)

**Comment by sksamuel** [comment]
lol

**Comment by helmbold** [comment]
I've just discovered that 22 is the one true limit, even in Kotlin! ;-) 22 is the max parameter count of a function, whereas type parameters seems to be unlimited. Now I have remove Table23 and added a bunch of `forAll` methods. Table driven testing should be complete, except some documentation.

**Comment by helmbold** [comment]
@sksamuel Table-driven testing shows the shortcomings of inheritance. Spec classes have to extend `TableTesting` to have tables available. There is now way to use tables with all test styles. I think we should try to find alternatives to class inheritance (maybe interfaces + delegation).

**Comment by sksamuel** [comment]
TableTesting could just be an interface which all the specs inherit.

**Comment by sksamuel** [comment]
Yes but I'm going to change that, so that table tests are not a testing style, but an alternative to matchers. Ie,
```kotlin
"you use whatever testing style you like" {
  forAll(table,.....)
}
```

```kotlin
should("you use whatever testing style you like") {
  forAll(table,.....)
}
```

**Comment by sksamuel** [comment]
@helmbold can we indent with 2 spaces. I'm not sure what the "kotlin" way is, but I think 2 is far better than 4 once you use it (Scala style) because otherwise nested functions end up on monitor 2.

**Comment by sksamuel** [comment]
6c7d31f559bacd513853b0683f344244a9f41b1b

**Comment by helmbold** [comment]
I just got used to this Kotlin [coding convention](http://kotlinlang.org/docs/reference/coding-conventions.html), but I can live with 2 spaces.

**Comment by sksamuel** [comment]
Yeah they say 4, but it really is better with 2 (the Scala way) IMO.
Especially as you nest more stuff.

**Comment by helmbold** [comment]
I was a 2 space advocate for a long time, too. I've only used 4 spaces because of the Kotlin convention. Somehow I learned to like 4 spaces, but for deep nesting 2 spaces are better. So let's go with 2 spaces.

**Comment by sksamuel** [comment]
I want to do the same thing for property tests - that is make them spec agnostic. It means changing the way they work a bit, but we've never documented properties as a feature. They're in 1.2.0 but hidden

**Comment by helmbold** [comment]
I like your effort! And, yes, nothing was documented, so probably hardly anyone used it.

**Comment by sksamuel** [comment]
@russel might be

**Comment by russel** [comment]
Russel uses nightlies or compiling from development repositories because if it is released it is too old. :-)

The only thing you need to do to stop me moaning is to make sure I know the need to change and what to do. As a side effect you get feedback on how good I think the change is!

**Comment by sksamuel** [comment]
In 1.3 @russel you need to nest the property tests inside a normal spec.

@helmbold I think we're looking ok for a 1.3 release ?

**Comment by helmbold** [comment]
Yes, I think my part for  1.3 is done.
